### PR TITLE
fix: clear `RunnableQueue`s on `ApiJobQueue.stop()`

### LINKED ISSUE #1420: Client retries failed jobs after logging out/in without restart
## Description

Tested against 2.1.0 server with:
- 0.5.1
- #1415 (9414760317596c488c41f1497937a9b9151ab033)

## Steps to Reproduce
1. Create a new source
2. Log into the client and wait for it to sync
3. Send a reply and wait for it to send (optional, but this makes the issue very visible)
4. Disconnect from the network
5. Send a reply
6. Click "Sign out"
7. Reconnect
8. Sign in and wait for sync

## Expected Behavior

- On step 6, reply transitions to "Failed to send"
- On step 8, reply stays in "Failed to send" (a reply failure is considered ultimate for now)

## Actual Behavior

- On step 6, reply transitions to "Failed to send" as expected
- On step 8, reply sends.

## Logs

https://gist.github.com/eloquence/b502702021bf1bfd10099fb778ba3859 (0.5.1)

### PR DESCRIPTION
**Pull request by cfm**
# Description

Fixes #1420 by:

1. defining a highest-priority (0) `ClearQueueJob`; and
2. enqueuing a `ClearQueueJob` in `ApiJobQueue.stop()`; thereby
3. clearing the `ApiJobQueue`'s `RunnableQueue`s and stopping their processing threads non-blocking, without `QThread.wait()`ing for them to return. 

# Test Plan

1. [ ] Follow #1420's steps to reproduce on `main` and confirm the actual versus expected behavior.
2. [ ] Follow #1420's steps to reproduce on `1420-stale-jobs` and observe the expected behavior.
    - [ ] While the `SendReplyJob` is timing out, the GUI does not freeze.

# Checklist

## After merge

- [ ] Update ["How Queue Processing Works"](https://github.com/freedomofpress/securedrop-client/wiki/Architecture#how-queue-processing-works)
- [ ] Update ["Network Testing" QA stories](https://github.com/freedomofpress/securedrop-client/wiki/QA#network-testing)

## Before merge

If these changes modify code paths involving cryptography, the opening of files in VMs or network (via the RPC service) traffic, Qubes testing in the staging environment is required. For fine tuning of the graphical user interface, testing in any environment in Qubes is required. Please check as applicable:

 - [x] I have tested these changes in the appropriate Qubes environment
 - I do not have an appropriate Qubes OS workstation set up (the reviewer will need to test these changes)
 - These changes should not need testing in Qubes

If these changes add or remove files other than client code, the AppArmor profile may need to be updated. Please check as applicable:

 - I have updated the [AppArmor profile](https://github.com/freedomofpress/securedrop-client/blob/HEAD/files/usr.bin.securedrop-client)
 - [x] No update to the AppArmor profile is required for these changes
 - I don't know and would appreciate guidance

If these changes modify the database schema, you should include a database migration. Please check as applicable:

 - I have written a migration and upgraded a test database based on `main` and confirmed that the migration applies cleanly
 - I have written a migration but have not upgraded a test database based on `main` and would like the reviewer to do so
 - I need help writing a database migration
 - [x] No database schema changes are needed

### CODE REVIEW

**Comment by cfm** [comment]
After my trivial fixup in 1536403 yesterday, I spent this afternoon reproducing and investigating the remaining test flake in `GpgHelper.decrypt_submission_or_reply()`:

<details>
<summary><a href="https://app.circleci.com/pipelines/github/freedomofpress/securedrop-client/1750/workflows/22d5691e-6901-4d04-8c28-f205d66d7a01/jobs/7614">In <code>test-buster</a></code></summary>

```
python -m pytest -v --random-order-bucket global tests/functional/test_receive_message.py
============================= test session starts ==============================
platform linux -- Python 3.7.3, pytest-6.2.5, py-1.11.0, pluggy-1.0.0 -- /root/project/.venv/bin/python
cachedir: .pytest_cache
Using --random-order-bucket=global
Using --random-order-seed=702021

PyQt5 5.11.3 -- Qt runtime 5.11.2 -- Qt compiled 5.11.2
rootdir: /root/project, configfile: pytest.ini
plugins: flaky-3.7.0, vcr-1.0.2, random-order-1.0.4, qt-4.0.2, mock-3.6.1, forked-1.4.0, cov-3.0.0, xdist-2.5.0
collected 1 item                                                               

tests/functional/test_receive_message.py::test_receive_message_from_source PASSED [100%]
===Flaky Test Report===

test_receive_message_from_source passed 1 out of the required 1 times. Success!

===End Flaky Test Report===

============================== 1 passed in 3.39s ===============================
terminate called without an active exception
Fatal Python error: Aborted

Thread 0x00007f9e1d7fa700 (most recent call first):
  File "/usr/lib/python3.7/threading.py", line 296 in wait
  File "/usr/lib/python3.7/threading.py", line 331 in wait_for
  File "/root/project/securedrop_client/queue.py", line 169 in process

Current thread 0x00007f9e1dffb700 (most recent call first):
  File "/usr/lib/python3.7/subprocess.py", line 1582 in _try_wait
  File "/usr/lib/python3.7/subprocess.py", line 1624 in _wait
  File "/usr/lib/python3.7/subprocess.py", line 990 in wait
  File "/usr/lib/python3.7/subprocess.py", line 325 in call
  File "/root/project/securedrop_client/crypto.py", line 108 in decrypt_submission_or_reply
  File "/root/project/securedrop_client/api_jobs/downloads.py", line 279 in call_decrypt
  File "/root/project/securedrop_client/api_jobs/downloads.py", line 184 in _decrypt
  File "/root/project/securedrop_client/api_jobs/downloads.py", line 137 in call_api
  File "/root/project/securedrop_client/api_jobs/base.py", line 73 in _do_call_api
  File "/root/project/securedrop_client/queue.py", line 180 in process

Thread 0x00007f9e717d9740 (most recent call first):
Aborted (core dumped)
```

</details>
<details>
<summary>Locally (with <code>-X faulthandler</code>)</summary>

```
python -X faulthandler -m pytest -v --random-order-bucket global tests/functional/test_seen.py
============================= test session starts ==============================
platform linux -- Python 3.7.3, pytest-7.0.1, pluggy-1.0.0 -- /home/user/securedrop-client/.venv/bin/python
cachedir: .pytest_cache
PyQt5 5.11.3 -- Qt runtime 5.11.2 -- Qt compiled 5.11.2
Using --random-order-bucket=global
Using --random-order-seed=688211

rootdir: /home/user/securedrop-client, configfile: pytest.ini
plugins: vcr-1.0.2, cov-3.0.0, mock-3.7.0, forked-1.4.0, flaky-3.7.0, xdist-2.5.0, qt-4.0.2, random-order-1.0.4
collected 2 items                                                              

tests/functional/test_seen.py::test_seen_and_unseen PASSED               [ 50%]
tests/functional/test_seen.py::test_unseen_source_becomes_seen_on_click PASSED [100%]
===Flaky Test Report===

test_seen_and_unseen passed 1 out of the required 1 times. Success!
test_unseen_source_becomes_seen_on_click passed 1 out of the required 1 times. Success!

===End Flaky Test Report===

============================== 2 passed in 5.99s ===============================
terminate called without an active exception
Fatal Python error: Aborted

Thread 0x00007edf6bfff700 (most recent call first):
  File "/usr/lib/python3.7/threading.py", line 296 in wait
  File "/usr/lib/python3.7/threading.py", line 331 in wait_for
  File "/home/user/securedrop-client/securedrop_client/queue.py", line 169 in process

Current thread 0x00007edf88cf3700 (most recent call first):
  File "/usr/lib/python3.7/subprocess.py", line 1582 in _try_wait
  File "/usr/lib/python3.7/subprocess.py", line 1624 in _wait
  File "/usr/lib/python3.7/subprocess.py", line 990 in wait
  File "/usr/lib/python3.7/subprocess.py", line 325 in call
  File "/home/user/securedrop-client/securedrop_client/crypto.py", line 108 in decrypt_submission_or_reply
  File "/home/user/securedrop-client/securedrop_client/api_jobs/downloads.py", line 279 in call_decrypt
  File "/home/user/securedrop-client/securedrop_client/api_jobs/downloads.py", line 184 in _decrypt
  File "/home/user/securedrop-client/securedrop_client/api_jobs/downloads.py", line 137 in call_api
  File "/home/user/securedrop-client/securedrop_client/api_jobs/base.py", line 73 in _do_call_api
  File "/home/user/securedrop-client/securedrop_client/queue.py", line 180 in process

Thread 0x00007edf8ad68700 (most recent call first):
  File "/usr/lib/python3.7/threading.py", line 296 in wait
  File "/usr/lib/python3.7/threading.py", line 331 in wait_for
  File "/home/user/securedrop-client/securedrop_client/queue.py", line 169 in process

Thread 0x00007edf8b569700 (most recent call first):
  File "/usr/lib/python3.7/threading.py", line 296 in wait
  File "/usr/lib/python3.7/threading.py", line 331 in wait_for
  File "/home/user/securedrop-client/securedrop_client/queue.py", line 169 in process

Thread 0x00007edfb1aef740 (most recent call first):
  File "/usr/lib/python3.7/weakref.py", line 358 in remove
./test-functional.sh: line 8:  8170 Aborted                 "${TEST_CMD[@]}"
```
</details>

Tl;dr:  The *tests* pass; the *test-handler* fails when its exit interrupts a blocking call in one of our threads.  The failure-mode *sounds* related to [`test-functional.sh`'s one test per `pytest` invocation](https://github.com/freedomofpress/securedrop-client/blob/f5f6b94ebeaa7d80475eb103714e49dd7eb7b37f/test-functional.sh#L3-L5) (discussed in #1010) but is pretty obviously a casualty of the thread-handling changes proposed here.  (Neither @creviera nor I have been able to reproduce this failure-mode in interactive testing on this branch.)

Here it's in [`subprocess.call()` in `GpgHelper.decrypt_submission_or_reply()`](https://github.com/freedomofpress/securedrop-client/blob/817c264b5dcc4b4d7aff57c21b542a05e2951ffe/securedrop_client/crypto.py#L108).  If I [refactor that `subprocess.call()` into a `QProcess`](https://github.com/cfm/securedrop-client/commit/4ff2ee1f0a1bdfa8a96b45230e7494861dcefd02), connected to the `QApplication.aboutToQuit` signal, the problem resurfaces in our SQLAlchemy session.

I'd be grateful for team feedback on:
    
1. whether there's anything obviously amiss in the minimal changes in this branch that accounts for this behavior; and
2. whether we've appetite for taking this as our cue to look into #1010 and improve how our functional tests interact with our `QApplication` and its `QThreads`; or
3. whether we'd rather shelve this approach to #1420 in favor of a [higher-level one](https://github.com/freedomofpress/securedrop-client/issues/1420#issuecomment-1054911116) for now.

**Comment by cfm** [comment]
<https://github.com/freedomofpress/securedrop-client/issues/1420#issuecomment-1057607965> points to a possible root cause for this failure-mode.

**Comment by cfm** [comment]
[Most recently](https://app.circleci.com/pipelines/github/freedomofpress/securedrop-client/1760/workflows/8cdc072d-eba3-4557-9333-d6992fc1033b/jobs/7640/steps), this crash occurs in `QProcess.waitForFinish()` despite having a slot connected to the `QApplication.aboutToQuit` signal:
    
https://github.com/freedomofpress/securedrop-client/blob/5e87a251333a6bd76c2f86b64834ec831efe215e/securedrop_client/crypto.py#L110-L116    
    
This makes me think that the answer to the operative question here, "Is this an *application*-shutdown problem or a *test*-teardown problem?", may well be the latter—for discussion with the team next week.

**Comment by cfm** [comment]
This is now ready for review at f735088e2aa4585aabdebade6d4eadfc291c92b2, focusing *just* on hooking `RunnableQueue.clear()` into `ApiJobQueue.stop()` after each `RunnableQueue`'s processing thread has stopped.  I'll follow up separately with some observations about why this was challenging to implement safely in the current queue-processing architecture.

File: securedrop_client/queue.py
Comment (creviera) [inline securedrop_client/queue.py:93]: :open_mouth: 

good find

**Review by creviera — APPROVED** [review]
- [x] Follow #1420's steps to reproduce on `main` and confirm the actual versus expected behavior.
- [x] Follow #1420's steps to reproduce on `1420-stale-jobs` and observe the expected behavior.

:tada: :tada: :tada: IT WORKS! :)

**Comment by creviera** [comment]
### STR

This is the simplest STR I could come up with.

1. Log in
2. From the Tor Control Panel, Stop Tor
3. Send a reply (it seems to be necessary to have a job in the queue)
4. Log out

### Expected

Immediate logout

### Actual

The client window freeze, you are not logged out and you can't exit the window

**Comment by cfm** [comment]
@creviera in <https://github.com/freedomofpress/securedrop-client/pull/1434#issuecomment-1063617133>:

> The client window freeze, you are not logged out and you can't exit the window

Ah, yup, because `Controller.logout()` is waiting for return from `ApiJobQueue.stop()`, which is now blocking on `QThread.wait()`:

https://github.com/freedomofpress/securedrop-client/blob/c1fde99c904c690420f1da958b39b0a98d52988a/securedrop_client/logic.py#L776-L778

I agree that we want logout to be an immediate action, even if the underlying queue mechanics (now) have to be blocking.  Come Monday, I'll:

1. **test this hypothesis by just moving up the `self.gui.logout()` call:**  logout should be immediate, and *then* the app will freeze pending `ApiJobQueue.stop()`;
2. **if confirmed, experiment with whether solving this calls for a signal on `Controller` or on `ApiJobQueue`.**  Design opinions welcome.

**Comment by creviera** [comment]
> > The client window freeze, you are not logged out and you can't exit the window
> 
> Ah, yup, because `Controller.logout()` is waiting for return from `ApiJobQueue.stop()`, which is now blocking on `QThread.wait()`

What surprised me was that it froze instead of taking X seconds to stop the queue and then complete the logout. It seems like `QThread.wait` could be waiting for more than just stopping the queue, since that should be pretty fast, right? It might have to do with the processing loop waits until something is in the queue, see https://github.com/freedomofpress/securedrop-client/blob/c1fde99c904c690420f1da958b39b0a98d52988a/securedrop_client/queue.py#L155. I haven't tested this hypothesis yet.

**Comment by cfm** [comment]
@creviera in <https://github.com/freedomofpress/securedrop-client/pull/1434#issuecomment-1067160277>:

> What surprised me was that it froze instead of taking X seconds to stop the queue and then complete the logout. It seems like `QThread.wait` could be waiting for more than just stopping the queue, since that should be pretty fast, right? It might have to do with the processing loop waiting until something is in the queue, see
> 
> https://github.com/freedomofpress/securedrop-client/blob/c1fde99c904c690420f1da958b39b0a98d52988a/securedrop_client/queue.py#L155
> Seems like that wait could cause a freeze. I haven't tested this hypothesis yet.

<details>
<summary>In my testing on Thursday, it's waiting for the <code>SendReplyJob</code> (as the current/next job in the queue) to time out with <code>SendReplyJobTimeoutError</code>.</summary>

```
Mar 10 12:14:07 localhost 2022-03-10 12:14:07,039 - securedrop_client.queue:185(process) DEBUG: SendReplyJobTimeoutError: Failed to send reply for source 8a288d23-556a-4821-be67-f0ec84e83373 due to Exception: Cannot connect to the server.
Mar 10 12:14:07 localhost 2022-03-10 12:14:07,040 - securedrop_client.queue:121(add_job) DEBUG: Added <securedrop_client.api_jobs.base.PauseQueueJob object at 0x74071c47ef78> to queue
Mar 10 12:14:07 localhost 2022-03-10 12:14:07,041 - securedrop_client.queue:138(_re_add_job) DEBUG: Added SendReplyJob('114f058e-1029-4337-8249-c386d1ec22e3', 0) to queue
Mar 10 12:14:07 localhost 2022-03-10 12:14:07,042 - securedrop_client.queue:255(stop) DEBUG: Stopped main queue
Mar 10 12:14:07 localhost 2022-03-10 12:14:07,043 - securedrop_client.queue:258(stop) DEBUG: Cleared main queue
Mar 10 12:14:07 localhost 2022-03-10 12:14:07,044 - securedrop_client.queue:121(add_job) DEBUG: Added <securedrop_client.api_jobs.base.PauseQueueJob object at 0x74071c47ef78> to queue
Mar 10 12:14:07 localhost 2022-03-10 12:14:07,046 - securedrop_client.queue:264(stop) DEBUG: Stopped file download queue
Mar 10 12:14:07 localhost 2022-03-10 12:14:07,047 - securedrop_client.queue:267(stop) DEBUG: Cleared file download queue
Mar 10 12:14:07 localhost 2022-03-10 12:14:07,102 - securedrop_client.logic:1129(on_reply_failure) DEBUG: 114f058e-1029-4337-8249-c386d1ec22e3 failed to send
Mar 10 12:14:07 localhost 2022-03-10 12:14:07,103 - securedrop_client.queue:274(on_main_queue_paused) DEBUG: Paused main queue
Mar 10 12:14:07 localhost 2022-03-10 12:14:07,104 - securedrop_client.queue:274(on_main_queue_paused) DEBUG: Paused main queue
Mar 10 12:14:07 localhost 2022-03-10 12:14:07,105 - securedrop_client.logic:487(completed_api_call) DEBUG: Completed API call. Cleaning up and running callback.
Mar 10 12:14:07 localhost 2022-03-10 12:14:07,105 - root:1144(on_logout_failure) INFO: Client logout failure
```
</summary>

**Comment by creviera** [comment]
It seems to also freeze if the sync (happening on a different thread) is in progress: 
Peek 2022-03-14 12-13

**Comment by cfm** [comment]
<https://github.com/freedomofpress/securedrop-client/pull/1434#issuecomment-1064482299>:

> Ah, yup, because `Controller.logout()` is waiting for return from `ApiJobQueue.stop()`, which is now blocking on `QThread.wait()`:
> 
> https://github.com/freedomofpress/securedrop-client/blob/c1fde99c904c690420f1da958b39b0a98d52988a/securedrop_client/logic.py#L776-L778
> 
> I agree that we want logout to be an immediate action, even if the underlying queue mechanics (now) have to be blocking. Come Monday, I'll:
> 
> 1. **test this hypothesis by just moving up the `self.gui.logout()` call:**  logout should be immediate, and _then_ the app will freeze pending `ApiJobQueue.stop()`;

I can confirm this with:

```diff
--- a/securedrop_client/logic.py
+++ b/securedrop_client/logic.py
@@ -773,9 +773,11 @@ class Controller(QObject):
         for failed_reply in failed_replies:
             self.reply_failed.emit(failed_reply.uuid)
 
+        self.gui.logout()
+        logger.debug('Moved GUI to logout state pending cleanup...')
+
         self.api_sync.stop()
         self.api_job_queue.stop()
-        self.gui.logout()
 
         self.show_last_sync_timer.start(TIME_BETWEEN_SHOWING_LAST_SYNC_MS)
         self.show_last_sync()
```

```
Mar 14 12:30:55 localhost 2022-03-14 12:30:55,724 - securedrop_client.logic:776(logout) DEBUG: Moved GUI to logout state pending cleanup...
Mar 14 12:30:55 localhost 2022-03-14 12:30:55,787 - securedrop_client.sync:69(stop) DEBUG: Stopping sync thread
Mar 14 12:30:55 localhost 2022-03-14 12:30:55,788 - securedrop_client.queue:121(add_job) DEBUG: Added <securedrop_client.api_jobs.base.PauseQueueJob object at 0x73798c229948> to queue
Mar 14 12:30:56 localhost 2022-03-14 12:30:56,463 - securedrop_client.logic:126(call_api) ERROR: Cannot connect to the server.
Mar 14 12:30:57 localhost 2022-03-14 12:30:57,724 - securedrop_client.queue:185(process) DEBUG: SendReplyJobTimeoutError: Failed to send reply for source 8a288d23-556a-4821-be67-f0ec84e83373 due to Exception: Cannot connect to the server.
Mar 14 12:30:57 localhost 2022-03-14 12:30:57,725 - securedrop_client.queue:121(add_job) DEBUG: Added <securedrop_client.api_jobs.base.PauseQueueJob object at 0x73798c1af678> to queue
Mar 14 12:30:57 localhost 2022-03-14 12:30:57,726 - securedrop_client.queue:138(_re_add_job) DEBUG: Added SendReplyJob('e95319a2-d1b6-4a16-9e56-f9cd42788396', 0) to queue
Mar 14 12:30:57 localhost 2022-03-14 12:30:57,727 - securedrop_client.queue:255(stop) DEBUG: Stopped main queue
Mar 14 12:30:57 localhost 2022-03-14 12:30:57,728 - securedrop_client.queue:258(stop) DEBUG: Cleared main queue
Mar 14 12:30:57 localhost 2022-03-14 12:30:57,728 - securedrop_client.queue:121(add_job) DEBUG: Added <securedrop_client.api_jobs.base.PauseQueueJob object at 0x73798c229948> to queue
Mar 14 12:30:57 localhost 2022-03-14 12:30:57,729 - securedrop_client.queue:264(stop) DEBUG: Stopped file download queue
Mar 14 12:30:57 localhost 2022-03-14 12:30:57,729 - securedrop_client.queue:267(stop) DEBUG: Cleared file download queue
Mar 14 12:30:57 localhost 2022-03-14 12:30:57,752 - securedrop_client.logic:487(completed_api_call) DEBUG: Completed API call. Cleaning up and running callback.
Mar 14 12:30:57 localhost 2022-03-14 12:30:57,753 - root:1147(on_logout_failure) INFO: Client logout failure
```

@creviera, I'm not sure if what you're seeing in <https://github.com/freedomofpress/securedrop-client/pull/1434#issuecomment-1067191814> is coming from `ApiSync.stop()`, which is unmodified in this PR and remains non-blocking on `QThread.quit()` without `.wait()`, or a parallel side effect of the blocking `ApiJobQueue.stop()`.
 
> 2. **if confirmed, experiment with whether solving this calls for a signal on `Controller` or on `ApiJobQueue`.**  Design opinions welcome.

I'll solicit your and @rocodes's design recommendations for this, based on your Qt expertise, this afternoon.  :-)

**Comment by creviera** [comment]
> @creviera, I'm not sure if what you're seeing in [#1434 (comment)](https://github.com/freedomofpress/securedrop-client/pull/1434#issuecomment-1067191814) is coming from `ApiSync.stop()`, which is unmodified in this PR and remains non-blocking on `QThread.quit()` without `.wait()`, or a parallel side effect of the blocking `ApiJobQueue.stop()`.
 
This issue is only introduced with this PR branch. The STR (which I'll add to subsequent tests of this PR) is:
1. wait for a sync to begin (you'll see the animation flashing)
2. logout
Expected: the client logs you out
Actual: the client freezes

You can also try:
1. send a message from the SI
2. wait for the sync to show the status update "Retrieving new messages..."
3. logout
Expected: the client logs you out
Actual: the client freezes
 
^ The second STR makes the most sense right off the bat because it enqueues download (and decrypt) jobs for new messages into the main queue. But since the solution we discussed in person today would no longer block the GUI when the user logs out, this issue will go away.

**Comment by cfm** [comment]
@creviera in <https://github.com/freedomofpress/securedrop-client/pull/1434#issuecomment-1067426492>:

> ^ The second STR makes the most sense right off the bat because it enqueues download (and decrypt) jobs for new messages into the main queue. But since the solution we discussed in person today would no longer block the GUI when the user logs out, this issue will go away.

Yes, I agree.  What's blocking in f735088e2aa4585aabdebade6d4eadfc291c92b2 must be not the apparent `ApiSync.stop()` call, then, but the `ApiJobQueue.stop()`, blocking on `QThread.wait()` for the `ApiJobQueue.download_file_queue` jobs to time out.

So, as we discussed this afternoon, here's 48a271980acfd24dda360d6f34ad20ff65b490c1 (newly rebased from `main`), which uses a `ClearQueueJob` to keep `ApiJobQueue.stop()` non-blocking on `QThread.stop()` *without* `QThread.wait()`.

<details><summary>Here's what this looks like in practice.</summary>

```
Mar 14 18:09:49 localhost 2022-03-14 18:09:49,491 - securedrop_client.queue:125(add_job) DEBUG: Added SendReplyJob('96ca0bb5-7c97-465c-818a-fd992a08c1b9', 5) to queue
[...]
Mar 14 18:09:53 localhost 2022-03-14 18:09:53,845 - securedrop_client.sync:69(stop) DEBUG: Stopping sync thread
Mar 14 18:09:53 localhost 2022-03-14 18:09:53,845 - securedrop_client.queue:125(add_job) DEBUG: Added <securedrop_client.api_jobs.base.ClearQueueJob object at 0x769bac667d38> to queue
Mar 14 18:09:53 localhost 2022-03-14 18:09:53,845 - securedrop_client.queue:271(stop) DEBUG: Asked main queue thread to quit
Mar 14 18:09:53 localhost 2022-03-14 18:09:53,846 - securedrop_client.queue:125(add_job) DEBUG: Added <securedrop_client.api_jobs.base.ClearQueueJob object at 0x769bac667dc8> to queue
Mar 14 18:09:53 localhost 2022-03-14 18:09:53,846 - securedrop_client.queue:276(stop) DEBUG: Asked file-download queue thread to quit
Mar 14 18:09:54 localhost 2022-03-14 18:09:54,012 - securedrop_client.queue:307(on_file_download_queue_cleared) DEBUG: Cleared file download queue
[...]
Mar 14 18:09:54 localhost 2022-03-14 18:09:54,757 - securedrop_client.logic:126(call_api) ERROR: Cannot connect to the server.
Mar 14 18:09:54 localhost 2022-03-14 18:09:54,758 - securedrop_client.logic:487(completed_api_call) DEBUG: Completed API call. Cleaning up and running callback.
Mar 14 18:09:54 localhost 2022-03-14 18:09:54,758 - root:1144(on_logout_failure) INFO: Client logout failure
[...]
Mar 14 18:09:54 localhost 2022-03-14 18:09:54,903 - securedrop_client.queue:195(process) DEBUG: SendReplyJobTimeoutError: Failed to send reply for source 8a288d23-556a-4821-be67-f0ec84e83373 due to Exception: Cannot connect to the server.
Mar 14 18:09:54 localhost 2022-03-14 18:09:54,905 - securedrop_client.queue:125(add_job) DEBUG: Added <securedrop_client.api_jobs.base.PauseQueueJob object at 0x769bac667ca8> to queue
Mar 14 18:09:54 localhost 2022-03-14 18:09:54,907 - securedrop_client.queue:142(_re_add_job) DEBUG: Added SendReplyJob('96ca0bb5-7c97-465c-818a-fd992a08c1b9', 0) to queue
[...]
Mar 14 18:09:55 localhost 2022-03-14 18:09:55,018 - securedrop_client.logic:1129(on_reply_failure) DEBUG: 96ca0bb5-7c97-465c-818a-fd992a08c1b9 failed to send
Mar 14 18:09:55 localhost 2022-03-14 18:09:55,018 - securedrop_client.queue:299(on_main_queue_cleared) DEBUG: Cleared main queue
```
</details>

As we discussed out of band, this allows things to happen *properly* out of order:

1. Sync starts to fail without network connectivity.
2. A `SendReplyJob` is enqueued.
3. The user initiates logout, which (a) enqueues a `ClearQueueJob` and (b) moves the GUI to the logout state.
4. The logout API call fails.
5. The `SendReplyJob` fails with `SendReplyJobTimeout`—*and is reenqueued!*
6. Instead, the higher-priority `ClearQueueJob` is processed next and clears the queue.

The current test failure is a mypy false positive around the typing of `{Clear,Pause}QueueJob`s, which are `QueueJob`s but *not* `ApiJobs`.  I'll sort that out tomorrow, pending any other feedback you may have!

**Comment by creviera** [comment]
> 1. Sync starts to fail without network connectivity.
> 2. A `SendReplyJob` is enqueued.
> 3. The user initiates logout, which (a) enqueues a `ClearQueueJob` and (b) moves the GUI to the logout state.
> 4. The logout API call fails.
> 5. The `SendReplyJob` fails with `SendReplyJobTimeout`—_and is reenqueued!_

Yup! And re-enqueuing is something that we may want to stop doing in the future, but that's the way it currently works so that we can auto-resend the reply as soon as the client can reach the server again and it unpauses the queue.

> 6. Instead, the higher-priority `ClearQueueJob` is processed next and clears the queue.

I think that captures it. There's still an edge case where the user quickly logs back in again, before the `ClearQueueJob` is processed, and quickly starts sending replies, etc. All the jobs that are enqueued before the `ClearQueueJob` is processed will be dropped. This is really unlikely, and I think the proper way to handle it will be to use a logout modal that closes the client once the `ClearQueueJob` completes, which can be done outside of this PR.

**Comment by cfm** [comment]
@creviera, this is ready for re-review per <https://github.com/freedomofpress/securedrop-client/pull/1434#issuecomment-1067463632>!  I've updated the test plan to include a check that <https://github.com/freedomofpress/securedrop-client/pull/1434#issuecomment-1063617133> has been resolved.

**Comment by creviera** [comment]
@cfm I believe this will also resolve https://github.com/freedomofpress/securedrop-client/issues/1444, correct?

**Comment by creviera** [comment]
:+1:  In the meantime, I've created a draft PR to address #1444, which will be on hold until this PR is merged.

So, currently, I'm seeing unexpected behavior where you have to log in twice when following the STR in the PR issue. I can give this a shot myself, but I think the `api_client` needs to be set to None upon logout so that we don't have to wait until login before we get an expired token. That's my best guess atm...

**Comment by cfm** [comment]
@creviera and I met today to discuss the following complications from this proposal as of e00e243, which I'm partially able to reproduce by varying the network failure-mode:

| Undesired behavior | After `sd-app` is detached from `sd-whonix`[1] | After Tor is stopped in `sd-whonix` |
| --- | --- | --- |
| 1. Stale (e.g.) `SendReplyJob` processed after logout/login | ✗, couldn't reproduce | ✓, reproduced |
| 2. Logout/login requires second login attempt | ✗ | ✗ |
| 3. Sync light flashes while still offline after `Controller.on_sync_failure` fires | ✗ | ✓ |

**Notes:**

1. Per original reproduction in #1420.
2. Original bug behavior reported in #1420.

We don't like the architectural implications (read: compromises), at multiple layers of the application, of any of the in-place workarounds we were able to come up with on the spot.  So we're moving this proposal back to draft status while we look further into what the original #1420 and these complications here reveal about how we're managing the Client's syncs, queues, and sessions.  If we wind up returning to this approach, we'll revisit this pull request; otherwise, we'll close it in favor of what we come up with instead.

**Comment by cfm** [comment]
In consultation with (especially) @eloquence and @creviera, I've added this back to the board for an implementation spike in the upcoming sprint:

Can we make logout a blocking/modal operation (as proposed in the course of #1452) by *killing* the sync and queue threads immediately?  Is this any worse than just quitting the app as currently implemented?
- What job-level operations do we need for this? interrupt/abort? don't retry? don't run?
- What queue-level operations do we need for this? stop? kill? clear?
- Is this different for read versus write jobs?

**Comment by cfm** [comment]
Today's findings from beginning the implementation spike proposed above:

1. Changing `ApiJobQueue.stop()` from `QThread.quit()` to `QThread.terminate()` does indeed prevent this call from blocking (freezing) the GUI.
2. However, this does not prevent the current job from completing (and succeeding if the network becomes available again).

<details><summary>See logs:</summary>

```
Mar 30 16:54:44 localhost 2022-03-30 16:54:44,810 - securedrop_client.queue:310(resume_queues) DEBUG: Resuming main queue
Mar 30 16:54:44 localhost 2022-03-30 16:54:44,810 - securedrop_client.queue:313(resume_queues) DEBUG: Resuming download queue
Mar 30 16:54:55 localhost 2022-03-30 16:54:55,404 - securedrop_client.queue:126(add_job) DEBUG: Added SendReplyJob('230d53e3-5ea2-4271-83c3-8e2226d76a36', 5) to queue
Mar 30 16:54:55 localhost 2022-03-30 16:54:55,411 - securedrop_client.crypto:159(import_key) DEBUG: Importing key for source d1573140-e174-4752-8ac6-7c4ec97e339c
Mar 30 16:55:00 localhost 2022-03-30 16:55:00,450 - securedrop_client.sync:69(stop) DEBUG: Stopping sync thread
Mar 30 16:55:00 localhost 2022-03-30 16:55:00,450 - securedrop_client.queue:265(stop) DEBUG: Terminated main queue
Mar 30 16:55:00 localhost 2022-03-30 16:55:00,451 - securedrop_client.queue:294(on_main_queue_cleared) DEBUG: Cleared main queue
Mar 30 16:55:00 localhost 2022-03-30 16:55:00,451 - securedrop_client.queue:270(stop) DEBUG: Terminated file download queue
Mar 30 16:55:00 localhost 2022-03-30 16:55:00,451 - securedrop_client.queue:302(on_file_download_queue_cleared) DEBUG: Cleared file download queue
Mar 30 16:55:21 localhost 2022-03-30 16:55:21,034 - securedrop_client.logic:1124(on_reply_success) INFO: 230d53e3-5ea2-4271-83c3-8e2226d76a36 sent successfully
Mar 30 16:55:21 localhost 2022-03-30 16:55:21,301 - securedrop_client.logic:487(completed_api_call) DEBUG: Completed API call. Cleaning up and running callback.
Mar 30 16:55:21 localhost 2022-03-30 16:55:21,301 - root:1144(on_logout_success) INFO: Client logout successful
```
</details>

This finding suggests that the priority in this spike should be isolating and solving why `ApiJobQueue.stop()` (in any implementation) does not either interrupt or fail the job currently under processing.  That's architecturally illuminating as well.  Thanks, @eloquence, for suggesting zooming back in with this spike!

**Comment by cfm** [comment]
Closing as a red herring per <https://github.com/freedomofpress/securedrop-client/issues/1420#issuecomment-1123063139>.

**Pull request closed without merging**
